### PR TITLE
Attempt at OpenElm

### LINKED ISSUE #6868: Support for OpenELM of Apple
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [ ] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [ ] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [ ] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [ ] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Feature Description

Support for OpenELM of Apple

https://huggingface.co/apple/OpenELM-3B-Instruct/tree/main

### PR DESCRIPTION
**Pull request by joshcarp**
Currently failing [on line 821 of sgemm.cpp](https://github.com/ggerganov/llama.cpp/compare/master...joshcarp:llama.cpp:openelm-impl?expand=1#diff-b89d92ffb96d84a060ec2a3ce66cd939edf2c8af1d3bc9edc133040bc5947eb1R821), still some parsing of ffn/attention head info needs to occur. Currently hard coded some stuff. 

Fixes: https://github.com/ggerganov/llama.cpp/issues/6868

Raising this PR as a draft because I need help. Will be adding comments to original source for reference purposes.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -10610,6 +10709,178 @@ struct llm_build_context {
...
+
+                Kcur = ggml_rope_custom(
+                    ctx0, Kcur, inp_pos, n_rot, rope_type, 0, n_orig_ctx,
+                    freq_base, freq_scale, ext_factor, attn_factor, beta_fast, beta_slow
+                );
+
+                // So because our original wo matrix wasn't 3x, the below function fails because there aren't enough elems in it.
+                // Got: [head_dim][n_tokens][n_head_v]
+                // Want: [n_embd_v_gqa(384)][n_tokens]
+                // I guess this means that i need to be able to able to repeat them
```
Comment (joshcarp) [inline llama.cpp:10800]: This is the problem, these things need to be repeated like they are in the python part on line 10806

Comment (joshcarp) [reply llama.cpp:10800]:
```python
if self.num_groups != 1:
                        # GQA
                        # [B, k_h, S, h] --> [B, q_h, S, h] // so, k=3 -> q=12
                        keys = keys.repeat_interleave(self.num_groups, dim=1)
                        # [B, v_h, S, h] --> [B, q_h, S, h] // so, v=3 -> q=12
```

Comment (joshcarp) [reply llama.cpp:10800]: trying to do something similar to this

Comment (joshcarp) [reply llama.cpp:10800]: I think this is solved now

File: llama.cpp [hunk 2]
```diff
@@ -10610,6 +10709,178 @@ struct llm_build_context {
...
+            cb(cur, "ffn_norm", il);
+            // FF
+            {
+
+                cur = llm_build_norm(ctx0, cur, hparams,
+                model.layers[il].ffn_norm, NULL,
+                LLM_NORM_RMS, cb, il);
+                cb(cur, "ffn_norm", il);
+
+                cur = llm_build_ffn(ctx0, cur,
```
Comment (joshcarp) [inline llama.cpp:10856]: Now i think i just need to implement `LLM_FFN_SWIGLU` which is slightly different to the existing activation functions i think:
image

**Comment by joshcarp** [comment]
Okay so this does the build purely because the tensor sizes all match up, which doesn't necesarily mean that it's the correct implementation shown by the fact that there's still a sigsev somewhere

**Comment by joshcarp** [comment]
Debugging this because some of the transformations are obvs wrong:

```
ggml_debug:                 inp_embd = (f32)   GET_ROWS(token_embd.weight{1280, 32000, 1, 1}, inp_tokens{3, 1, 1, 1}}) = {1280, 3, 1, 1}
                                     sum = 138204.000000
ggml_debug:                   norm-0 = (f32)   RMS_NORM(inp_embd{1280, 3, 1, 1}, }) = {1280, 3, 1, 1}
                                     sum = 138204.015625
ggml_debug:              attn_norm-0 = (f32)        MUL(norm-0{1280, 3, 1, 1}, blk.0.attn_norm.weight{1280, 1, 1, 1}}) = {1280, 3, 1, 1}
                                     sum = 138204.000000
ggml_debug:                   wqkv-0 = (f32)    MUL_MAT(blk.0.attn_qkv.weight{1280, 1152, 1, 1}, attn_norm-0{1280, 3, 1, 1}}) = {1152, 3, 1, 1}
                                     sum = 124380.968750
ggml_debug:            wqkv-0 (view) = (f32)       VIEW(wqkv-0{1152, 3, 1, 1}, }) = {64, 3, 12, 1}
                                     sum = 8722732.000000
ggml_debug:     wqkv-0 (view) (cont) = (f32)       CONT(wqkv-0 (view){64, 3, 12, 1}, }) = {64, 3, 12, 1}
                                     sum = 497453.906250
ggml_debug:                   norm-0 = (f32)   RMS_NORM(wqkv-0 (view) (cont){64, 3, 12, 1}, }) = {64, 3, 12, 1}
                                     sum = 497532.062500
ggml_debug:                   Qcur-0 = (f32)        MUL(norm-0{64, 3, 12, 1}, blk.0.attn_q_norm.weight{64, 1, 1, 1}}) = {64, 3, 12, 1}
                                     sum = 497535.562500
ggml_debug:        Qcur-0 (reshaped) = (f32)    RESHAPE(Qcur-0{64, 3, 12, 1}, }) = {64, 12, 3, 1}
                                     sum = 497535.593750
ggml_debug:                   Qcur-0 = (f32)       ROPE(Qcur-0 (reshaped){64, 12, 3, 1}, inp_pos{3, 1, 1, 1}}) = {64, 12, 3, 1}
                                     sum = 497535.593750
ggml_debug:                   Qcur-0 = (f32)      SCALE(Qcur-0{64, 12, 3, 1}, }) = {64, 12, 3, 1}
                                     sum = 497458.562500
ggml_debug:            wqkv-0 (view) = (f32)       VIEW(wqkv-0{1152, 3, 1, 1}, }) = {64, 3, 3, 1}
                                     sum = 1002128.312500
ggml_debug:     wqkv-0 (view) (cont) = (f32)       CONT(wqkv-0 (view){64, 3, 3, 1}, }) = {64, 3, 3, 1}
                                     sum = 62096.066406
ggml_debug:                   norm-0 = (f32)   RMS_NORM(wqkv-0 (view) (cont){64, 3, 3, 1}, }) = {64, 3, 3, 1}
                                     sum = 62041.996094
ggml_debug:                   ''Kcur-0 = (f32)        MUL(norm-0{64, 3, 3, 1}, blk.0.attn_k_norm.weight{64, 1, 1, 1}}) = {64, 3, 3, 1}
                                     sum = 62035.175781
ggml_debug:        Kcur-0 (reshaped) = (f32)    RESHAPE(Kcur-0{64, 3, 3, 1}, }) = {64, 3, 3, 1}
                                     sum = 62035.175781
ggml_debug:                  node_16 = (f32)       ROPE(Kcur-0 (reshaped){64, 3, 3, 1}, inp_pos{3, 1, 1, 1}}) = {64, 3, 3, 1}
                                     sum = 62035.175781
ggml_debug:                  node_17 = (f32)     REPEAT(node_16{64, 3, 3, 1}, }) = {64, 3, 12, 1}
                                     sum = 497318.656250
ggml_debug:            wqkv-0 (view) = (f32)       VIEW(wqkv-0{1152, 3, 1, 1}, }) = {64, 3, 3, 1}
                                     sum = 1002132.000000
ggml_debug:                   Vcur-0 = (f32)       CONT(wqkv-0 (view){64, 3, 3, 1}, }) = {64, 3, 3, 1}
                                     sum = 62100.000000
ggml_debug:                  node_20 = (f32)     REPEAT(Vcur-0{64, 3, 3, 1}, }) = {64, 3, 12, 1}
                                     sum = 497448.000000
ggml_debug:                      v-0 = (f16)       VIEW(cache_v_l0{655360, 1, 1, 1}, }) = {32, 64, 6, 1}
                                     sum = nan
```

**Comment by uhlhosting** [comment]
How is progress on this?

**Comment by hunt-47** [comment]
mahn, we all ae waiting for you...1 
@joshcarp nice work by the way..

**Comment by joshcarp** [comment]
Still slaving away at debugging, 
Like i said before, first time in this codebase so things are a bit wack.
All of the parameter and tensor parsing/shaping i'm pretty certain is correct, i'm comparing against the reference implementation through corenet+huggingface

**Comment by joshcarp** [comment]
Closing because i'm not making any meaningful progress on this

**Pull request closed without merging**